### DISCUSSION TITLE
How can i pass pathname and query in Next js 13.4 to router.push() in App router?

### DISCUSSION THREAD

**Discussion by Raghunadh9** [vercel/next.js#55906] [App Router]
My instructor using Next js 11 version, he used to `import router from "next/router"`. but as we know that `next/router` has been depreciated.  he passes pathname and query to the `router.push` like this:
```
 router.push({
      pathname: path,
      query: query,
    });
```
How can i do it in App router? We need to update. not degrade. any suggestions?

**Comment by geekerio** [comment]
you can use 

import { useRouter} from 'next/navigation';

const router = useRouter()
router.push(`/path?query=${query}`)

**Reply by palminha** [reply]
@geekerio that's an ugly way... we should be able to do with `route/navigation` the same as we did with `route/router`, particularly with `pathname` and `query`

@Raghunadh9 I'm struggling as you... how did you solve it?

**Reply by Aman-Clement** [reply]
@geekerio I'm struggling with the same, did you find any workaround?

**Reply by Devintel-byte** [reply]
I had same issues and tried some workarounds, this solution worked for me:

import {useRouter} from "next/navigation"

const navigate = useRouter();

const query = new URLSearchParams({
imageSrc,
title,
postSponsor,
}).toString();

navigate.push(`/donation?${query}`);

the imageSrc, title, postSponsor are params.

**Comment by Devintel-byte** [comment]
I had same issues and tried some workarounds, this solution worked for me:

import {useRouter} from "next/navigation"

const navigate = useRouter();

const query = new URLSearchParams({
      imageSrc,
      title,
      postSponsor,
    }).toString();

    navigate.push(`/donation?${query}`);
    
    the imageSrc, title, postSponsor are params.

**Comment by Farzin-Firoozi** [comment]
You can write a custom hook to achieve it, here's an example I use in my projects:
```tsx
import { useRouter as useNextRouter } from 'next/navigation'

const useRouter = () => {
  const { push: nextPush, ...rest } = useNextRouter()

  const customPush = ({
    pathname,
    query,
  }: {
    pathname: string
    query: Record<string, any>
  }) => {
    const generateQuery = new URLSearchParams(query).toString()

    nextPush(`${pathname}?${generateQuery}`)
  }

  return { ...rest, push: customPush }
}

export default useRouter
```
